### PR TITLE
Implement select aggregate commit in commit pool - Closes #6846

### LINKED ISSUE #6846: Implement selectAggregateCommit of CommitPool
### Description

- Implement `selectAggregateCommit()` as per specified in [LIP:Commit:selectAggregateCommit](https://research.lisk.com/t/introduce-certificate-generation-mechanism/296#selectaggregatecommit-45) 
```ts
public getAggregageCommit() {
  return  this._aggregateSingleCommits(this._selectAggregateCommit());
};
```
### Acceptance Criteria

- Add unit test for `selectAggregateCommit`

### PR DESCRIPTION
**Pull request by mitsuaki-u**
### What was the problem?

This PR resolves #6846 

### How was it solved?

- 🌱 Implement selectAggregateCommit a9ade2bffdf30b5f675bb50e13fce9386a9816e0
- 🌱 Added tests d9d6c638f7a6c8b287daeebb79293999222dd51b

### How was it tested?

Adding unit tests

### CODE REVIEW

File: framework/src/node/consensus/certificate_generation/commit_pool.ts [hunk 1]
```diff
@@ -169,17 +169,73 @@ export class CommitPool {
...
+		let heightNextBFTParameters: number;
+		let nextHeight: number;
+
+		try {
+			heightNextBFTParameters = await this._bftAPI.getNextHeightBFTParameters(
+				apiContext,
+				maxHeightCertified + 1,
+			);
+			nextHeight = Math.min(heightNextBFTParameters - 1, maxHeightPrecommitted);
+		} catch (err) {
```
Comment (shuse2) [inline framework/src/node/consensus/certificate_generation/commit_pool.ts:193]: this should differentiate `BFTParameterNotFoundError` error and other error

File: framework/src/node/consensus/certificate_generation/commit_pool.ts [hunk 2]
```diff
@@ -169,17 +169,73 @@ export class CommitPool {
...
+		let singleCommits: SingleCommit[];
+		let validatorsInPool: Buffer[];
+		let aggregateBFTWeight = BigInt(0);
+
+		while (nextHeight > maxHeightCertified) {
+			singleCommits = [
+				...(this._nonGossipedCommits.get(nextHeight) ?? []),
+				...(this._gossipedCommits.get(nextHeight) ?? []),
+			];
+			validatorsInPool = singleCommits.map(commit => commit.validatorAddress);
```
Comment (shuse2) [inline framework/src/node/consensus/certificate_generation/commit_pool.ts:206]: `nextValidators`?

File: framework/src/node/consensus/certificate_generation/commit_pool.ts [hunk 3]
```diff
@@ -169,17 +169,73 @@ export class CommitPool {
...
+
+			try {
```
Comment (shuse2) [inline framework/src/node/consensus/certificate_generation/commit_pool.ts:208]: i think you can assume `BFTParameters` exists for the next height

File: framework/src/node/consensus/certificate_generation/commit_pool.ts [hunk 4]
```diff
@@ -169,17 +169,73 @@ export class CommitPool {
...
+				maxHeightCertified + 1,
+			);
+			nextHeight = Math.min(heightNextBFTParameters - 1, maxHeightPrecommitted);
+		} catch (err) {
+			nextHeight = maxHeightPrecommitted;
+		}
+
+		let singleCommits: SingleCommit[];
+		let validatorsInPool: Buffer[];
+		let aggregateBFTWeight = BigInt(0);
```
Comment (shuse2) [inline framework/src/node/consensus/certificate_generation/commit_pool.ts:199]: these values should be declared in the `while` scope

File: framework/src/node/consensus/certificate_generation/commit_pool.ts [hunk 5]
```diff
@@ -169,17 +169,73 @@ export class CommitPool {
...
+						throw new Error('Validator address not found in commit pool');
+					}
+
+					aggregateBFTWeight += bftParamsValidatorInfo.bftWeight;
+
+					if (aggregateBFTWeight >= certificateThreshold) {
+						return this._aggregateSingleCommits(singleCommits);
+					}
+
+					nextHeight -= 1;
```
Comment (shuse2) [inline framework/src/node/consensus/certificate_generation/commit_pool.ts:228]: this should be outside of for loop?

**Review by shuse2 — CHANGES_REQUESTED** [review]
I will review the test after these changes are addressed

File: framework/src/node/consensus/certificate_generation/commit_pool.ts [hunk 6]
```diff
@@ -168,18 +168,71 @@ export class CommitPool {
...
+				);
+				if (!bftParamsValidatorInfo) {
+					throw new Error('Validator address not found in commit pool');
+				}
+
+				aggregateBFTWeight += bftParamsValidatorInfo.bftWeight;
+
+				if (aggregateBFTWeight >= certificateThreshold) {
+					return this._aggregateSingleCommits(singleCommits);
+				}
```
Comment (shuse2) [inline framework/src/node/consensus/certificate_generation/commit_pool.ts:226]: this block should happen outside of for loop, and if it is lower than the certificate threshold, it needs to check `nextHeight -= 1`

File: framework/test/unit/node/consensus/certificate_generation/commit_pool.spec.ts [hunk 7]
```diff
@@ -335,7 +336,149 @@ describe('CommitPool', () => {
...
+			const result = await commitPool['_selectAggregateCommit'](apiContext);
+
+			// Assert
+			expect(commitPool['_aggregateSingleCommits']).not.toHaveBeenCalled();
+			expect(result).toEqual({
+				height: maxHeightCertified,
+				aggregationBits: Buffer.alloc(0),
+				certificateSignature: Buffer.alloc(0),
+			});
+		});
```
Comment (shuse2) [inline framework/test/unit/node/consensus/certificate_generation/commit_pool.spec.ts:482]: these tests need to be re-reviewed with the src change

**Review by shuse2 — CHANGES_REQUESTED** [review]

**Review by shuse2 — APPROVED** [review]
🚀

File: framework/test/unit/node/consensus/certificate_generation/commit_pool.spec.ts (hunk 7 above)
Comment (mitsuaki-u) [reply framework/test/unit/node/consensus/certificate_generation/commit_pool.spec.ts:482]: the tests are still valid as the previous code flow actually detected when `aggregateSingleCommits` reached threshold sooner. with the new code flow it detects it after the for loop. either way the test is still valid.

File: framework/src/node/consensus/certificate_generation/commit_pool.ts [hunk 8]
```diff
@@ -168,18 +168,71 @@ export class CommitPool {
...
+				return this._aggregateSingleCommits(singleCommits);
+			}
+
+			nextHeight -= 1;
+		}
+
+		return {
+			height: maxHeightCertified,
+			aggregationBits: Buffer.alloc(0),
+			certificateSignature: Buffer.alloc(0),
```
Comment (ishantiw) [inline framework/src/node/consensus/certificate_generation/commit_pool.ts:235]: we could define `export const EMPTY_BUFFER = Buffer.alloc(0);` in constants and use that instead

**Review by ishantiw — APPROVED** [review]
Looks good 🚀 , just a minor comment to consider

**Pull request merged**